nomad/periodic.go

```diff
@@ -0,0 +1,529 @@
+		p.logger.Printf("[DEBUG] nomad.periodic: job %q has no valid launch time", nextJob.job.ID)
+		select {
+		case <-p.stopCh:
+			return
+		case <-p.updateCh:
+			goto PICK
+		}
+	}
+
+	now = time.Now()
+	p.logger.Printf("[DEBUG] nomad.periodic: launching job %q in %s",
+		nextJob.job.ID, nextJob.next.Sub(now))
+
+	select {
+	case <-p.stopCh:
+		return
+	case <-p.updateCh:
+		goto PICK
+	case <-time.After(nextJob.next.Sub(now)):
+		// Get the current time so that we don't miss any jobs will we are
+		// creating evals.
+		nowUpdate := time.Now()
+
+		// Create evals for all the jobs with the same launch time.
+		p.l.Lock()
+		for {
+			if p.heap.Length() == 0 {
+				break
+			}
+
+			j, err := p.heap.Peek()
+			if err != nil {
+				p.logger.Printf("[ERR] nomad.periodic: failed to determine next periodic job: %v", err)
+				break
+			}
+
+			if j.next != launchTime {
+				break
+			}
+
+			if err := p.heap.Update(j.job, j.job.Periodic.Next(nowUpdate)); err != nil {
+				p.logger.Printf("[ERR] nomad.periodic: failed to update next launch of periodic job %q: %v", j.job.ID, err)
+			}
+
+			p.logger.Printf("[DEBUG] nomad.periodic: launching job %v at %v", j.job.ID, launchTime)
+			go p.createEval(j.job, launchTime)
+		}
+
+		p.l.Unlock()
+		now = nowUpdate
+	}
+
+	goto PICK
+}
+
+// createEval instantiates a job based on the passed periodic job and submits an
+// evaluation for it.
+func (p *PeriodicDispatch) createEval(periodicJob *structs.Job, time time.Time) error {
+	derived, err := p.deriveJob(periodicJob, time)
+	if err != nil {
+		return err
+	}
+
+	// Commit this update via Raft
+	req := structs.JobRegisterRequest{Job: derived}
+	_, index, err := p.srv.raftApply(structs.JobRegisterRequestType, req)
+	if err != nil {
+		p.logger.Printf("[ERR] nomad.periodic: registering child job for periodic job %q failed: %v", periodicJob.ID, err)
+		return err
+	}
+
+	// Create a new evaluation
+	eval := &structs.Evaluation{
+		ID:             structs.GenerateUUID(),
+		Priority:       derived.Priority,
+		Type:           derived.Type,
+		TriggeredBy:    structs.EvalTriggerJobRegister,
+		JobID:          derived.ID,
+		JobModifyIndex: index,
+		Status:         structs.EvalStatusPending,
+	}
+	update := &structs.EvalUpdateRequest{
+		Evals: []*structs.Evaluation{eval},
+	}
+
+	// Commit this evaluation via Raft
+	// XXX: There is a risk of partial failure where the JobRegister succeeds
+	// but that the EvalUpdate does not.
+	_, _, err = p.srv.raftApply(structs.EvalUpdateRequestType, update)
+	if err != nil {
+		p.logger.Printf("[ERR] nomad.periodic: creating eval for %q failed: %v", derived.ID, err)
+		return err
+	}
+
+	return nil
+}
+
+// deriveJob instantiates a new job based on the passed periodic job and the
+// launch time.
+// TODO: these jobs need to be marked as GC'able
+func (p *PeriodicDispatch) deriveJob(periodicJob *structs.Job, time time.Time) (
+	derived *structs.Job, err error) {
+
+	// Have to recover in case the job copy panics.
```

Comment:
Why is this possible? We can do an explicit copy of all the modified fields instead of depending on copy structure if this is an issue.